nixos/modules/services/databases/neo4j.nix

```diff
@@ -36,114 +89,567 @@ let
+              directory is mandatory and will need to be created manually when:
+              setting this option to something other than its default; setting
+              either this policy's <option>baseDirectory</option> or
+              <option>directories.certificates</option> to something other than
+              their default. Ensure read/write permissions are given to the
+              Neo4j daemon user <literal>neo4j</literal>.
+              </para>
+              <para>
+              The public certificate as defined by
+              <option>publicCertificate</option> is required to be duplicated
+              to this directory.
+            '';
+          };
+
+        };
+      }));
+      default = {};
+      description = ''
+        Defines the SSL policies for use with Neo4j connectors. Each attribute
+        of this set defines a policy, with the attribute name defining the name
+        of the policy and its namespace. Refer to the operations manual section
+        on Neo4j's
+        <link xlink:href="https://neo4j.com/docs/operations-manual/current/security/ssl-framework/">SSL Framework</link>
+        for further details.
+      '';
     };
 
-    extraServerConfig = mkOption {
-      description = "Extra configuration for neo4j server.";
-      default = "";
-      type = types.lines;
+    udc = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enable the Usage Data Collector which Neo4j uses to collect usage
+          data. Refer to the operations manual section on the
+          <link xlink:href="https://neo4j.com/docs/operations-manual/current/configuration/usage-data-collector/">Usage Data Collector</link>
+          for more information.
+        '';
+      };
     };
+
   };
 
   ###### implementation
 
-  config = mkIf cfg.enable {
-    systemd.services.neo4j = {
-      description = "Neo4j Daemon";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
-      environment = {
-        NEO4J_HOME = "${cfg.package}/share/neo4j";
-        NEO4J_CONF = "${cfg.dataDir}/conf";
-      };
-      serviceConfig = {
-        ExecStart = "${cfg.package}/bin/neo4j console";
-        User = "neo4j";
-        PermissionsStartOnly = true;
-        LimitNOFILE = 40000;
-      };
-      preStart = ''
-        mkdir -m 0700 -p ${cfg.dataDir}/{data/graph.db,conf,logs}
-        ln -fs ${serverConfig} ${cfg.dataDir}/conf/neo4j.conf
-        if [ "$(id -u)" = 0 ]; then chown -R neo4j ${cfg.dataDir}; fi
-      '';
-    };
+  config =
+    let
+      policyNameList = (attrNames cfg.ssl.policies);
+      validPolicyNameList = ["legacy"] ++ policyNameList;
+      validPolicyNameString = concatStringsSep ", " validPolicyNameList;
+
+      defaultDirectoriesList = builtins.filter (dir: dir.highestPrio >= 1500) (builtins.attrValues options.services.neo4j.directories);
```

Comment:
@Infinisil  I'm not understanding something about the list of option values as generated here.
```
while evaluating anonymous function at /home/pattern/repos/nixpkgs/nixos/modules/services/databases/neo4j.nix:570:49, called from undefined position:
value is a list while a set was expected, at /home/pattern/repos/nixpkgs/nixos/modules/services/databases/neo4j.nix:570:54
```
I thought `builtins.attrValues options.services.neo4j.directories` would return a list of the sets defined in `options.services.neo4j.directories`

I think I'm on the right path, but am stuck.

Reply:
Ah I got it, this is a fix:
```diff
diff --git a/nixos/modules/services/databases/neo4j.nix b/nixos/modules/services/databases/neo4j.nix
index bdf6f44dd1d..50c386ce9b6 100644
--- a/nixos/modules/services/databases/neo4j.nix
+++ b/nixos/modules/services/databases/neo4j.nix
@@ -567,11 +567,11 @@ in {
       validPolicyNameList = ["legacy"] ++ policyNameList;
       validPolicyNameString = concatStringsSep ", " validPolicyNameList;
 
-      defaultDirectoriesList = builtins.filter (dir: dir.highestPrio >= 1500) (builtins.attrValues options.services.neo4j.directories);
+      defaultDirectoriesList = builtins.filter (dir: isOption dir && dir.highestPrio >= 1500) (builtins.attrValues options.services.neo4j.directories);
       certDirIsDefault = options.services.neo4j.directories.certificates.highestPrio >= 1500;
       defaultPoliciesList =
         if !certDirIsDefault then []
-        else builtins.filter (pol: pol.baseDirectory.highestPrio >= 1500) (builtins.attrValues options.services.neo4j.ssl.policies);
+        else builtins.filter (pol: isOption pol && pol.baseDirectory.highestPrio >= 1500) (builtins.attrValues options.services.neo4j.ssl.policies);
     in
 
     mkIf cfg.enable {
```

The reason for this is that `options.services.neo4j.directories` actually also includes a value for `_definedNames`, which is a list used by the module system probably, only just now realized this, sorry! The fix just makes sure that all values are options via `isOption`.

Reply:
Admittedly this is a bit ugly, but imo that's better than the alternative (explicitly repeating code for all 5 directories)

Reply:
`highestPrio` doesn't seem to exist on these sets.
```nix
defaultDirectoriesList = builtins.filter (
  dir: (builtins.trace dir dir).highestPrio >= 1500
) (builtins.filter (dir: isOption dir) (builtins.attrValues options.services.neo4j.directories));
```
```
trace: { _type = "option"; declarations = <CODE>; default = <CODE>; definitions = <CODE>; description = "Directory for storing ..."; files = <CODE>; isDefined = <CODE>; loc = <CODE>; options = <CODE>; type = <CODE>; value = <CODE>; }

...
attribute 'highestPrio' missing, at /home/pattern/repos/nixpkgs/nixos/modules/services/databases/neo4j.nix:574:14
```

Reply:
Cannot reproduce: https://gist.github.com/8f954a39abcdf0f9e4903057134a6f69

I'm using the latest commit from here

Reply:
@Infinisil This appears to be due to my building against a branch made from my local nixos-version (18.03 stable). Looks like these commits made re: highestPrio 4017fdcafdd4a1a2063b74bc46c744c85ffcc22e and 441796cb16e0b97c0ddcf947c6c90d86c1118b3f are found in nixos-unstable, but not 18.03. I will work to build agains unstable instead. I suppose this technique would then not be compatible for backport to 18.03?

Reply:
Updates won't get backported anyways.